File: cv/scripts/game_state_tracker.py
#!/usr/bin/env python3
"""Persistent Durak game state reconstructed from consecutive CV observations."""

from __future__ import annotations

from dataclasses import dataclass, field


TERMINAL_EVENTS = {
    "PLAYER_TAKES",
    "OPPONENT_TAKES",
    "PLAYER_DEFENDED",
    "OPPONENT_DEFENDED",
}


@dataclass(frozen=True)
class ObservedCard:
    card: str
    zone: str
    class_confidence: float = 1.0
    box_confidence: float = 1.0
    x: float = 0.0
    y: float = 0.0

    @property
    def reliable(self) -> bool:
        return self.class_confidence >= 0.65 and self.box_confidence >= 0.20


@dataclass(frozen=True)
class FrameObservation:
    cards: tuple[ObservedCard, ...] = ()
    events: tuple[str, ...] = ()
    deck_count: int | None = None


@dataclass
class TrackerUpdate:
    changes: list[str] = field(default_factory=list)


class GameStateTracker:
    """Remember known cards instead of re-recognizing an overlapped hand."""

    def __init__(self, minimum_initial_hand: int = 1) -> None:
        self.hand: set[str] = set()
        self.table: set[str] = set()
        self.table_attack: list[str] = []
        self.table_defense: list[str] = []
        self.discard: set[str] = set()
        self.opponent_known: set[str] = set()
        self.player_attacker: bool | None = None
        self.initialized = False
        self.active_events: set[str] = set()
        self.draw_candidates: dict[str, int] = {}
        self.table_candidates: dict[str, int] = {}
        self.hand_reconcile_pending = False
        self.hand_reconcile_age = 0
        self.hand_snapshot_candidate: frozenset[str] = frozenset()
        self.hand_snapshot_count = 0
        self.last_deck_count: int | None = None
        self.empty_deck_grace = 0
        self.last_event_kind: int | None = None
        self.last_event_cards: tuple[str, ...] = ()
        self.last_event_actor_self: bool | None = None
        self.minimum_initial_hand = minimum_initial_hand

    def _remember_event(
        self, kind: int, cards=(), actor_self: bool | None = None,
    ) -> None:
        """Store the most recent inferred move using durak_v3 MOVE_KINDS indices."""
        self.last_event_kind = kind
        self.last_event_cards = tuple(cards)
        self.last_event_actor_self = actor_self

    def observe(self, observation: FrameObservation) -> TrackerUpdate:
        update = TrackerUpdate()
        if observation.deck_count == 0 and self.last_deck_count not in {None, 0}:
            # The last cards have just been dealt. Give the refreshed hand a
            # few frames to settle before enforcing the no-more-draws rule.
            self.empty_deck_grace = 3
        elif observation.deck_count not in {None, 0}:
            self.empty_deck_grace = 0
        deck_empty_locked = observation.deck_count == 0 and self.empty_deck_grace == 0
        current_events = set(observation.events)
        new_events = current_events - self.active_events
        visible_hand = {
            item.card for item in observation.cards
            if item.zone == "hand" and item.reliable
        }
        raw_table_items = sorted(
            (item for item in observation.cards if item.zone == "table"),
            key=lambda item: (item.y, item.x),
        )
        moderate_table = {
            item.card for item in raw_table_items
            if item.class_confidence >= 0.40 and item.box_confidence >= 0.30
        }
        self.table_candidates = {
            name: self.table_candidates.get(name, 0) + 1
            for name in moderate_table
        }
        confirmed_table = {
            name for name, confirmations in self.table_candidates.items()
            if confirmations >= 2
        }
        table_items = [
            item for item in raw_table_items
            if item.reliable or item.card in confirmed_table
        ]
        table_evidence = {
            item.card: item
            for item in sorted(
                table_items,
                key=lambda item: item.class_confidence * item.box_confidence,
            )
        }
        visible_table = {item.card for item in table_items}
        # During the take/discard animation the old cards can remain visible
        # for several frames. Once the terminal event was processed, never add
        # those pixels back as a new table until the bubble disappears.
        repeated_terminal = bool(current_events & TERMINAL_EVENTS) and not bool(
            new_events & TERMINAL_EVENTS
        )
        if repeated_terminal:
            visible_table.clear()
            table_items.clear()
            self.table_candidates.clear()

        if not self.initialized and len(visible_hand) >= self.minimum_initial_hand:
            self.hand = set(visible_hand)
            self.initialized = True
            update.changes.append("initial hand: " + " ".join(sorted(self.hand)))

        # After a completed bout (except when we took the table), the game fans
        # our refreshed hand out again. Two identical snapshots are safer than
        # carrying missed played cards forever, and remove false remembered
        # cards that were created from overlapped corners.
        if self.hand_reconcile_pending:
            self.hand_reconcile_age += 1
            snapshot = frozenset(visible_hand)
            if snapshot and len(snapshot) <= 6:
                if snapshot == self.hand_snapshot_candidate:
                    self.hand_snapshot_count += 1
                else:
                    self.hand_snapshot_candidate = snapshot
                    self.hand_snapshot_count = 1
                if self.hand_snapshot_count >= 2:
                    before = set(self.hand)
                    proposed = set(snapshot)
                    # With an empty deck, an unseen card cannot be a draw. A
                    # changed class name is therefore a crop error, not a real
                    # hand replacement (for example QD -> JD on an overlapped
                    # face card). Keep the remembered identities in that case;
                    # actual plays are removed from the table observation.
                    if deck_empty_locked and not proposed <= before:
                        proposed = before
                    self.hand = proposed
                    removed = before - self.hand
                    added = self.hand - before
                    details = []
                    if removed:
                        details.append("removed " + " ".join(sorted(removed)))
                    if added:
                        details.append("added " + " ".join(sorted(added)))
                    if details:
                        update.changes.append("hand reconciled: " + "; ".join(details))
                    self.hand_reconcile_pending = False
                    self.hand_snapshot_candidate = frozenset()
                    self.hand_snapshot_count = 0
                    self.draw_candidates.clear()
                    if observation.deck_count == 0:
                        self.empty_deck_grace = 0
            if self.hand_reconcile_age >= 5:
                self.hand_reconcile_pending = False
                self.hand_snapshot_candidate = frozenset()
                self.hand_snapshot_count = 0

        old_hand = set(self.hand)
        new_table = visible_table - self.table
        missing_table = self.table - visible_table
        # If the visible number of physical cards grew by N but there are more
        # than N new class names, some previous card was merely reclassified.
        # Prefer the very common correction where rank stays the same and only
        # the suit changes (for example JS -> JC on a narrow corner crop).
        physical_growth = max(0, len(visible_table) - len(self.table))
        corrections_needed = max(0, len(new_table) - physical_growth)
        corrections = []
        for old in sorted(missing_table):
            if corrections_needed <= 0:
                break
            same_rank = next((new for new in sorted(new_table) if new[:-1] == old[:-1]), None)
            if same_rank:
                # Temporal confirmation lets a moderately confident new card
                # join the table, but it must not rewrite an already known
                # card identity. Suit corrections require one independently
                # reliable observation; otherwise keep the remembered card.
                evidence = table_evidence.get(same_rank)
                if evidence is not None and evidence.reliable:
                    corrections.append((old, same_rank))
                new_table.remove(same_rank)
                missing_table.remove(old)
                corrections_needed -= 1
        for old, new in corrections:
            self.table.remove(old)
            self.table.add(new)
            self.table_attack = [new if card == old else card for card in self.table_attack]
            self.table_defense = [new if card == old else card for card in self.table_defense]
            update.changes.append(f"table corrected: {old} -> {new}")
        if new_table:
            table_was_empty = not self.table
            own_cards = {name for name in new_table if name in old_hand}
            opponent_cards = new_table - own_cards
            # Public cards played by the opponent are no longer in the subset
            # of their hand that we know exactly.
            self.opponent_known.difference_update(opponent_cards)
            if self.player_attacker is None:
                # The first card of a bout identifies who opened it.
                first = next(item.card for item in table_items if item.card in new_table)
                self.player_attacker = first in own_cards
            accepted_new = set()
            for item in table_items:
                if item.card not in new_table:
                    continue
                played_by_player = item.card in own_cards
                is_attack = played_by_player == self.player_attacker
                # A matching-rank card played before any defense is a transfer:
                # it joins the attacks and swaps attacker/defender roles.
                is_transfer = (
                    not is_attack
                    and bool(self.table_attack)
                    and not self.table_defense
                    and item.card[:-1] == self.table_attack[0][:-1]
                )
                if is_transfer:
                    is_attack = True
                    self.player_attacker = not self.player_attacker
                if not is_attack and len(self.table_defense) >= len(self.table_attack):
                    update.changes.append(f"ignored impossible extra defense: {item.card}")
                    continue
                target = self.table_attack if is_attack else self.table_defense
                target.append(item.card)
                accepted_new.add(item.card)
                event_kind = (
                    6 if is_transfer else
                    0 if is_attack and table_was_empty else
                    1 if is_attack else 2
                )
                self._remember_event(event_kind, (item.card,), played_by_player)
                table_was_empty = False
            self.table.update(accepted_new)
            if accepted_new:
                update.changes.append("table added: " + " ".join(sorted(accepted_new)))

        # A known card appearing on the table must have left our hand. This
        # covers attacks, defenses, throw-ins, and transfers uniformly.
        played = self.table & self.hand
        if played:
            self.hand.difference_update(played)
            update.changes.append("played: " + " ".join(sorted(played)))

        # Newly visible cards in a small hand are draws. Cards received after
        # taking are added from the remembered table below, even if overlapped.
        draws_allowed = (
            not deck_empty_locked
            and (not self.table or self.hand_reconcile_pending)
        )
        possible_draws = (
            visible_hand - self.hand - self.table - self.discard
            if draws_allowed else set()
        )
        self.draw_candidates = {
            name: self.draw_candidates.get(name, 0) + 1
            for name in possible_draws
        }
        # A narrow exposed corner is occasionally localized as a second card
        # (for example 7C -> 6C). New hand cards therefore need to agree in two
        # consecutive frames before they enter persistent memory.
        drawn = {
            name for name, confirmations in self.draw_candidates.items()
            if confirmations >= 2
        }
        if self.initialized and drawn:
            self.hand.update(drawn)
            for name in drawn:
                self.draw_candidates.pop(name, None)
            update.changes.append("recognized draw: " + " ".join(sorted(drawn)))
            if observation.deck_count == 0:
                self.empty_deck_grace = 0

        for event in observation.events:
            if event not in new_events:
                continue
            if event == "PLAYER_TAKES":
                self.draw_candidates.clear()
                self.table_candidates.clear()
                self.hand_reconcile_pending = False
                previous_attacker = self.player_attacker
                taken = set(self.table)
                self.hand.update(taken)
                self.table.clear()
                self.table_attack.clear()
                self.table_defense.clear()
                self.player_attacker = False
                # The attacker finishes the take after the defender forfeits.
                self._remember_event(5, (), previous_attacker is True)
                update.changes.append("player takes: " + (" ".join(sorted(taken)) or "empty table"))
            elif event in {"OPPONENT_TAKES", "PLAYER_DEFENDED", "OPPONENT_DEFENDED"}:
                self.draw_candidates.clear()
                self.table_candidates.clear()
                discarded = set(self.table)
                if event == "OPPONENT_TAKES":
                    self.opponent_known.update(discarded)
                if event in {"PLAYER_DEFENDED", "OPPONENT_DEFENDED"}:
                    self.discard.update(discarded)
                previous_attacker = self.player_attacker
                self.table.clear()
                self.table_attack.clear()
                self.table_defense.clear()
                self.player_attacker = event in {"OPPONENT_TAKES", "PLAYER_DEFENDED"}
                if event == "OPPONENT_TAKES":
                    self._remember_event(5, (), previous_attacker is True)
                else:
                    self._remember_event(4, (), previous_attacker is True)
                self.hand_reconcile_pending = True
                self.hand_reconcile_age = 0
                self.hand_snapshot_candidate = frozenset()
                self.hand_snapshot_count = 0
                update.changes.append(f"{event.lower()}: " + (" ".join(sorted(discarded)) or "empty table"))
            elif event in {"PLAYER_STOPPED_THROWING", "OPPONENT_STOPPED_THROWING"}:
                update.changes.append(event.lower())
        self.active_events = current_events
        if observation.deck_count is not None:
            self.last_deck_count = observation.deck_count
        if self.empty_deck_grace > 0:
            self.empty_deck_grace -= 1
        return update

    def summary(self) -> str:
        hand = " ".join(sorted(self.hand)) or "-"
        table = " ".join(sorted(self.table)) or "-"
        role = {True: "attack", False: "defense", None: "unknown"}[self.player_attacker]
        known = " ".join(sorted(self.opponent_known)) or "-"
        return (
            f"hand=[{hand}] table=[{table}] discard={len(self.discard)} "
            f"opponent_known=[{known}] role={role}"
        )

File: android/app/src/main/java/ru/kartegnik/durakadvisor/GameStateTracker.java
package ru.kartegnik.durakadvisor;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Reconstructs public game state from a sequence of card detections. */
final class GameStateTracker {
    enum Zone {
        OPPONENT,
        TABLE,
        HAND,
    }

    static final class ObservedCard {
        final String card;
        final Zone zone;
        final float classConfidence;
        final float boxConfidence;
        final float centerX;
        final float centerY;

        ObservedCard(
                String card, Zone zone, float classConfidence, float boxConfidence,
                float centerX, float centerY) {
            this.card = card;
            this.zone = zone;
            this.classConfidence = classConfidence;
            this.boxConfidence = boxConfidence;
            this.centerX = centerX;
            this.centerY = centerY;
        }

        boolean reliable() {
            return classConfidence >= 0.65f && boxConfidence >= 0.30f;
        }
    }

    private final Set<String> hand = new LinkedHashSet<>();
    private final Set<String> table = new LinkedHashSet<>();
    private final List<String> attacks = new ArrayList<>();
    private final List<String> defenses = new ArrayList<>();
    private final Set<String> discard = new LinkedHashSet<>();
    private final Set<String> knownOpponent = new LinkedHashSet<>();
    private final Map<String, Integer> drawCandidates = new HashMap<>();
    private Set<String> lastVisibleHand = Set.of();
    private Set<String> initialCandidate = Set.of();
    private Set<String> refreshCandidate = Set.of();
    private int initialCandidateFrames;
    private int refreshCandidateFrames;
    private int refreshAge;
    private int absentTableFrames;
    private boolean initialized;
    private boolean awaitingRefreshedHand;
    private Boolean playerAttacker;
    private int deckCount = 24;
    private int opponentCount = 6;
    private int expectedHandCount;
    private Integer lastEventKind;
    private List<String> lastEventCards = List.of();
    private Boolean lastEventActorSelf;
    private String suggestedOpeningCard;

    void reset() {
        hand.clear();
        table.clear();
        attacks.clear();
        defenses.clear();
        discard.clear();
        knownOpponent.clear();
        drawCandidates.clear();
        lastVisibleHand = Set.of();
        initialCandidate = Set.of();
        refreshCandidate = Set.of();
        initialCandidateFrames = 0;
        refreshCandidateFrames = 0;
        refreshAge = 0;
        absentTableFrames = 0;
        initialized = false;
        awaitingRefreshedHand = false;
        playerAttacker = null;
        deckCount = 24;
        opponentCount = 6;
        expectedHandCount = 0;
        lastEventKind = null;
        lastEventCards = List.of();
        lastEventActorSelf = null;
        suggestedOpeningCard = null;
    }

    void observe(List<ObservedCard> observed) {
        Set<String> visibleHand = new LinkedHashSet<>();
        List<ObservedCard> visibleTableItems = new ArrayList<>();
        for (ObservedCard card : observed) {
            if (card.zone == Zone.HAND && card.reliable()) {
                visibleHand.add(card.card);
            } else if (card.zone == Zone.TABLE
                    && card.classConfidence >= 0.40f
                    && card.boxConfidence >= 0.20f) {
                visibleTableItems.add(card);
            }
        }
        visibleTableItems.sort(Comparator
                .comparingDouble((ObservedCard card) -> card.centerY)
                .thenComparingDouble(card -> card.centerX));
        Set<String> visibleTable = new LinkedHashSet<>();
        for (ObservedCard card : visibleTableItems) {
            visibleTable.add(card.card);
        }

        if (!initialized) {
            stabilizeInitialHand(visibleHand);
            if (!initialized) {
                lastVisibleHand = Set.copyOf(visibleHand);
                return;
            }
        }

        boolean handCountDecreased = !lastVisibleHand.isEmpty()
                && visibleHand.size() < lastVisibleHand.size();

        if (!table.isEmpty()) {
            Set<String> overlap = new HashSet<>(visibleTable);
            overlap.retainAll(table);
            if (visibleTable.isEmpty()) {
                absentTableFrames++;
                if (absentTableFrames >= 2) {
                    settleBout();
                }
            } else if (overlap.isEmpty()) {
                // The previous implementation waited for a second disjoint
                // frame. During that delay the overlay could recommend a move
                // for the previous bout while already displaying the new
                // table. A completely different non-empty table is stronger
                // evidence than a transient empty detector frame, so switch
                // bouts immediately and process the visible opening card.
                settleBout();
                absentTableFrames = 0;
            } else {
                absentTableFrames = 0;
            }
        }

        processNewTableCards(visibleTableItems, handCountDecreased);
        // Card identities are unique in a 36-card deck. Even if a table card
        // was already remembered before this frame, it can no longer remain
        // in our hand. Keeping this invariant prevents stale suggestions such
        // as transferring with the same card already visible on the table.
        hand.removeAll(table);
        observeHand(visibleHand);
        lastVisibleHand = Set.copyOf(visibleHand);
    }

    private void stabilizeInitialHand(Set<String> visibleHand) {
        if (visibleHand.isEmpty()) {
            initialCandidate = Set.of();
            initialCandidateFrames = 0;
            return;
        }
        if (visibleHand.equals(initialCandidate)) {
            initialCandidateFrames++;
        } else {
            initialCandidate = Set.copyOf(visibleHand);
            initialCandidateFrames = 1;
        }
        if (initialCandidateFrames >= 2) {
            hand.addAll(visibleHand);
            initialized = true;
        }
    }

    private void processNewTableCards(
            List<ObservedCard> visibleTableItems, boolean handCountDecreased) {
        Set<String> oldHand = new HashSet<>(hand);
        oldHand.addAll(lastVisibleHand);
        boolean tableWasEmpty = table.isEmpty();
        int unseenCards = 0;
        for (ObservedCard item : visibleTableItems) {
            unseenCards += table.contains(item.card) ? 0 : 1;
        }
        for (ObservedCard item : visibleTableItems) {
            if (table.contains(item.card)) {
                continue;
            }
            boolean playedByPlayer = oldHand.contains(item.card);
            if (tableWasEmpty && !playedByPlayer
                    && item.card.equals(suggestedOpeningCard)) {
                playedByPlayer = true;
            }
            if (tableWasEmpty && !playedByPlayer
                    && unseenCards == 1 && handCountDecreased) {
                // Fallback for a rare table misclassification: if exactly one
                // card appeared while our visible hand lost a card, we opened.
                playedByPlayer = true;
            }
            if (tableWasEmpty) {
                // Re-observe the opener every bout instead of trusting a role
                // carried through an animation that CV may have skipped.
                playerAttacker = playedByPlayer;
                suggestedOpeningCard = null;
            }
            boolean isAttack = playedByPlayer == playerAttacker;
            boolean transfer = !isAttack && !attacks.isEmpty() && defenses.isEmpty()
                    && rank(item.card).equals(rank(attacks.get(0)));
            if (transfer) {
                isAttack = true;
                playerAttacker = !playerAttacker;
            }
            if (!isAttack && defenses.size() >= attacks.size()) {
                continue;
            }
            if (isAttack) {
                attacks.add(item.card);
            } else {
                defenses.add(item.card);
            }
            table.add(item.card);
            if (playedByPlayer) {
                hand.remove(item.card);
            } else {
                knownOpponent.remove(item.card);
                opponentCount = Math.max(0, opponentCount - 1);
            }
            rememberEvent(
                    transfer ? 6 : isAttack && tableWasEmpty ? 0 : isAttack ? 1 : 2,
                    List.of(item.card), playedByPlayer);
            tableWasEmpty = false;
        }
    }

    private void observeHand(Set<String> visibleHand) {
        if (awaitingRefreshedHand) {
            refreshAge++;
            if (!visibleHand.isEmpty() && visibleHand.equals(refreshCandidate)) {
                refreshCandidateFrames++;
            } else {
                refreshCandidate = Set.copyOf(visibleHand);
                refreshCandidateFrames = visibleHand.isEmpty() ? 0 : 1;
            }
            boolean complete = visibleHand.size() >= expectedHandCount;
            if ((complete && refreshCandidateFrames >= 2) || refreshAge >= 6) {
                if (complete) {
                    hand.clear();
                }
                hand.addAll(visibleHand);
                awaitingRefreshedHand = false;
                refreshCandidate = Set.of();
                refreshCandidateFrames = 0;
                drawCandidates.clear();
            }
            return;
        }

        Set<String> possibleDraws = new HashSet<>(visibleHand);
        possibleDraws.removeAll(hand);
        possibleDraws.removeAll(table);
        possibleDraws.removeAll(discard);
        Map<String, Integer> updated = new HashMap<>();
        for (String card : possibleDraws) {
            int confirmations = drawCandidates.getOrDefault(card, 0) + 1;
            if (confirmations >= 2) {
                hand.add(card);
            } else {
                updated.put(card, confirmations);
            }
        }
        drawCandidates.clear();
        drawCandidates.putAll(updated);
    }

    private void settleBout() {
        boolean fullyDefended = !attacks.isEmpty() && defenses.size() >= attacks.size();
        boolean previousPlayerAttacker = Boolean.TRUE.equals(playerAttacker);
        if (fullyDefended) {
            discard.addAll(table);
            rememberEvent(4, List.of(), previousPlayerAttacker);
        } else {
            if (previousPlayerAttacker) {
                knownOpponent.addAll(table);
                opponentCount += table.size();
            } else {
                hand.addAll(table);
            }
            rememberEvent(5, List.of(), previousPlayerAttacker);
        }
        table.clear();
        attacks.clear();
        defenses.clear();
        absentTableFrames = 0;
        refillCounts(previousPlayerAttacker);
        // Until the next physical opening card appears, show only a
        // conditional suggestion. Its owner will establish the real role.
        playerAttacker = null;
        awaitingRefreshedHand = true;
        refreshAge = 0;
        refreshCandidate = Set.of();
        refreshCandidateFrames = 0;
    }

    private void refillCounts(boolean playerDrewFirst) {
        int ownDraw = Math.max(0, 6 - hand.size());
        int opponentDraw = Math.max(0, 6 - opponentCount);
        if (playerDrewFirst) {
            int actualOwn = Math.min(deckCount, ownDraw);
            deckCount -= actualOwn;
            expectedHandCount = hand.size() + actualOwn;
            int actualOpponent = Math.min(deckCount, opponentDraw);
            deckCount -= actualOpponent;
            opponentCount += actualOpponent;
        } else {
            int actualOpponent = Math.min(deckCount, opponentDraw);
            deckCount -= actualOpponent;
            opponentCount += actualOpponent;
            int actualOwn = Math.min(deckCount, ownDraw);
            deckCount -= actualOwn;
            expectedHandCount = hand.size() + actualOwn;
        }
    }

    private void rememberEvent(int kind, Collection<String> cards, Boolean actorSelf) {
        lastEventKind = kind;
        lastEventCards = List.copyOf(cards);
        lastEventActorSelf = actorSelf;
    }

    boolean initialized() {
        return initialized;
    }

    void rememberSuggestedOpening(String card) {
        suggestedOpeningCard = card;
    }

    boolean readyForAdvice() {
        return initialized && !awaitingRefreshedHand;
    }

    Set<String> hand() {
        return Set.copyOf(hand);
    }

    Set<String> table() {
        return Set.copyOf(table);
    }

    List<String> attacks() {
        return List.copyOf(attacks);
    }

    List<String> defenses() {
        return List.copyOf(defenses);
    }

    Set<String> discard() {
        return Set.copyOf(discard);
    }

    Set<String> knownOpponent() {
        return Set.copyOf(knownOpponent);
    }

    Boolean playerAttacker() {
        return playerAttacker;
    }

    int deckCount() {
        return deckCount;
    }

    int opponentCount() {
        return Math.max(knownOpponent.size(), opponentCount);
    }

    Integer lastEventKind() {
        return lastEventKind;
    }

    List<String> lastEventCards() {
        return lastEventCards;
    }

    Boolean lastEventActorSelf() {
        return lastEventActorSelf;
    }

    private static String rank(String card) {
        return card.substring(0, card.length() - 1);
    }
}
